jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormData.java

```diff
@@ -409,251 +182,445 @@ protected HttpFields customizePartHeaders(MultiPart.Part part)
+        {
+            return filesDirectory;
+        }
+
+        /**
+         * <p>Sets the directory where the files uploaded in the parts will be saved.</p>
+         *
+         * @param filesDirectory the directory where files are saved
+         */
+        public void setFilesDirectory(Path filesDirectory)
+        {
+            this.filesDirectory = filesDirectory;
+        }
+
+        /**
+         * @return the maximum file size in bytes, or -1 for unlimited file size
+         */
+        public long getMaxFileSize()
+        {
+            return maxFileSize;
+        }
+
+        /**
+         * @param maxFileSize the maximum file size in bytes, or -1 for unlimited file size
+         */
+        public void setMaxFileSize(long maxFileSize)
+        {
+            this.maxFileSize = maxFileSize;
+        }
+
+        /**
+         * @return the maximum memory file size in bytes, or -1 for unlimited memory file size
+         */
+        public long getMaxMemoryFileSize()
+        {
+            return maxMemoryFileSize;
+        }
+
+        /**
+         * <p>Sets the maximum memory file size in bytes, after which files will be saved
+         * in the directory specified by {@link #setFilesDirectory(Path)}.</p>
+         * <p>Use value {@code 0} to always save the files in the directory.</p>
+         * <p>Use value {@code -1} to never save the files in the directory.</p>
+         *
+         * @param maxMemoryFileSize the maximum memory file size in bytes, or -1 for unlimited memory file size
+         */
+        public void setMaxMemoryFileSize(long maxMemoryFileSize)
+        {
+            this.maxMemoryFileSize = maxMemoryFileSize;
+        }
+
+        /**
+         * @return the maximum length in bytes of the whole multipart content, or -1 for unlimited length
+         */
+        public long getMaxLength()
+        {
+            return maxLength;
+        }
+
+        /**
+         * @param maxLength the maximum length in bytes of the whole multipart content, or -1 for unlimited length
+         */
+        public void setMaxLength(long maxLength)
+        {
+            this.maxLength = maxLength;
+        }
+
+        /**
+         * @return the maximum number of parts that can be parsed from the multipart content.
+         */
+        public long getMaxParts()
+        {
+            return parser.getMaxParts();
+        }
+
+        /**
+         * @param maxParts the maximum number of parts that can be parsed from the multipart content.
+         */
+        public void setMaxParts(long maxParts)
+        {
+            parser.setMaxParts(maxParts);
+        }
+
+        // Only used for testing.
+        int getPartsSize()
+        {
+            return listener.getPartsSize();
+        }
+
+        private class PartsListener extends MultiPart.AbstractPartsListener
+        {
+            private final AutoLock lock = new AutoLock(); // TODO why do we need this lock?
```

Comment:
We need the lock because `fail()` can be called at any time asynchronously from the parsing.

Reply:
But can it be now?  Failures come via source.read().  If somebody externally fails the CompletableFuture, then the atomic nature of that helps with memory barriers and making things only happen once.   What exactly is the scenario that two threads can be in the parser listener at the same time?

Reply:
It is possible because we return the CF to the application, and they can fail it at any time, and failing the CF must cleanup what's been partially parsed.